src/connection.ts

```diff
@@ -1284,6 +1284,72 @@ export class Connection<S extends Schema = Schema> extends EventEmitter {
     records: Record | Record[],
     extIdField: string,
     options: DmlOptions = {},
+  ): Promise<SaveResult | SaveResult[]> {
+    return Array.isArray(records)
+    ? // check the version whether SObject collection API is supported (46.0)
+      this._ensureVersion(46)
+      ? this._upsertMany(type, records, extIdField, options)
+      : this._upsertParallel(type, records, extIdField, options)
+    : this._upsertParallel(type, records, extIdField, options);
+}
+
+/** @private */
+  async _upsertMany(
+    type: string,
+    records: Record | Record[],
+    extIdField: string,
+    options: DmlOptions = {},
+  ): Promise<SaveResult | SaveResult[]> {
+    if (records.length === 0) {
+      return [];
+    }
+    if (records.length > MAX_DML_COUNT && options.allowRecursive) {
+      return [
+        ...((await this._upsertMany(
+          type,
+          records.slice(0, MAX_DML_COUNT),
+          extIdField,
+          options,
+        )) as SaveResult[]),
+        ...((await this._upsertMany(type, records.slice(MAX_DML_COUNT), extIdField, options)) as SaveResult[]),
+      ];
+    }
+    const _records = records.map((recordItem: Record) => {
+      const { [extIdField]: extId, type: recordType, attributes, ...rec } = recordItem;
+      const sobjectType = recordType || attributes?.type || type;
+      if (!extId) {
+        throw new Error('External ID is not found in record.');
+      } 
+      if (!sobjectType) {
+        throw new Error('No SObject Type defined in record');
+      }
+      return { [extIdField]: extId, attributes: { type: sobjectType }, ...rec };
+    });
+    let url =
+      [this._baseUrl(), 'composite', 'sobjects', type, extIdField].join('/');
+    if (options.allOrNone) {
+      url += '&allOrNone=true';
+    }
```

Comment:
nit:
this URL param can be removed, only `DELETE` seems to respect it:
https://developer.salesforce.com/docs/atlas.en-us.api_rest.meta/api_rest/resources_composite_sobjects_collections_delete.htm

create/update/upsert take `allOrNone` in the request body.